core/src/main/java/com/linecorp/armeria/common/util/StartStopSupport.java

```diff
@@ -384,6 +395,11 @@ private void notifyListeners(State state, @Nullable T startArg, @Nullable U stop
                 }
             } catch (Exception cause) {
                 notificationFailed(l, cause);
+
+                // if notifyStarting throws an exception, propagate the exception
+                if (state == State.STARTING) {
+                    throw cause;
+                }
```

Comment:
How about wrapping the cause with `IllegalStateException("Failed to start the server", cause)` and removing `throws Exception`?